provider/source/src/units/categorized_display_name.rs

```diff
@@ -0,0 +1,153 @@
+                    .marker_attributes
+                    .split_once('-')
+                    .ok_or_else(|| DataErrorKind::InvalidRequest.into_error())?;
+
+                let units_format_data: &cldr_serde::units::data::Resource = self
+                    .cldr()?
+                    .units()
+                    .read_and_parse(req.id.locale, "units.json")?;
+                let units_format_data = &units_format_data.main.value.units;
+
+                let unit_patterns = match length {
+                    "long" => &units_format_data.long,
+                    "short" => &units_format_data.short,
+                    "narrow" => &units_format_data.narrow,
+                    _ => {
+                        return Err(DataErrorKind::InvalidRequest
+                            .into_error()
+                            .with_debug_context(length))
+                    }
+                }
+                .categories
+                .get(category)
+                .ok_or_else(|| {
+                    DataErrorKind::IdentifierNotFound
+                        .into_error()
+                        .with_debug_context(category)
+                })?
+                .get(unit)
+                .ok_or_else(|| {
+                    DataErrorKind::IdentifierNotFound
+                        .into_error()
+                        .with_debug_context(length)
+                })?;
+
+                Ok(DataResponse {
+                    metadata: Default::default(),
+                    payload: DataPayload::from_owned(UnitsDisplayName {
+                        patterns: PluralElements::new(
+                            unit_patterns
+                                .other
+                                .as_deref()
+                                .ok_or_else(|| DataErrorKind::IdentifierNotFound.into_error())?,
+                        )
+                        .with_zero_value(unit_patterns.zero.as_deref())
+                        .with_one_value(unit_patterns.one.as_deref())
+                        .with_two_value(unit_patterns.two.as_deref())
+                        .with_few_value(unit_patterns.few.as_deref())
+                        .with_many_value(unit_patterns.many.as_deref())
+                        .with_explicit_one_value(unit_patterns.explicit_one.as_deref())
+                        .with_explicit_zero_value(unit_patterns.explicit_zero.as_deref())
+                        .into(),
+                    }),
+                })
+            }
+        }
+    };
+}
+
+impl_categorized_display_name_data_provider!("area", AreaDisplayNameV1);
+impl_categorized_display_name_data_provider!("duration", DurationDisplayNameV1);
+impl_categorized_display_name_data_provider!("length", LengthDisplayNameV1);
+impl_categorized_display_name_data_provider!("mass", MassDisplayNameV1);
+impl_categorized_display_name_data_provider!("volume", VolumeDisplayNameV1);
+
+macro_rules! impl_categorized_display_name_iter_data_provider_cached {
```

Comment:
Nit: There is a lot of duplicated code between these two macros. Please try to reduce the duplication.